File: c++/OOP3Behav2.cpp
#include "pt4.h"
#include <list>
#include <iterator>
using namespace std;

class Observer;

class Subject
{
public:
    virtual void Attach(Observer* observ) = 0;
    virtual void Detach(Observer* observ) = 0;
    virtual void SetInfo(string info) = 0;
};

class Observer
{
public:
    virtual void OnInfo(Subject* sender, string info) = 0;
};

class ConcreteSubject : public Subject{
	list<Observer*> observers;
	list<Observer*>::iterator iterator;
	//string message;
	public:
		void Attach(Observer* observ) {
			observers.push_back(observ);
		};
    	void Detach(Observer* observ) {
    		//observers.remove(observ);
    		iterator = observers.erase(iterator);iterator --;
		};
    	void SetInfo(string info) {
			iterator = observers.begin();
			while (iterator != observers.end()) {
      			(*iterator)->OnInfo(this,info);
      			++iterator;}
		};
};

class ConcreteObserver : public Observer
{
	string log;
	char detachInfo;
public:
	ConcreteObserver (char detachInfo){
		this->detachInfo = detachInfo;
		log = "";
	}
    void OnInfo(Subject* sender, string info) {
    	log += info;
    	if (info[1] == detachInfo){
		Show("det");Show(info); Detach(sender);
		} 
	};
	void Attach (Subject* s){
		s->Attach(this);
	}
	void Detach (Subject* s){
		//Show("det");
		s->Detach(this);
	}
	string GetLog() {return log;}
};


// ÐÅÀËÈÇÓÉÒÅ ÊËÀÑÑÛ-ÏÎÒÎÌÊÈ ConcreteSubject
//   È ConcreteObserver

void Solve()
{
    Task("OOP3Behav2");
    
    ConcreteSubject* cs1 = new ConcreteSubject();
    ConcreteSubject* cs2 = new ConcreteSubject();
    
    int n,k;
    pt >> n >> k;
    
    ConcreteObserver* obs[n];
    
    for (int i=0;i<n;i++){
    	Show("init");
    	obs[i] = new ConcreteObserver(i+'a');
    	cs1->Attach(obs[i]);
    	cs2->Attach(obs[i]);
	}
	
	for (int i=0;i<k;i++){
		Show("in work");
		string tmp; pt >> tmp;
		if (tmp[0] == '1') cs1->SetInfo(tmp);
		else cs2->SetInfo(tmp);
	}
    
    for (int i=0;i<n;i++){
    	pt << obs[i]->GetLog();
	}
    
    

}

File: c#/OOP3Behav2.cs
// File: "OOP3Behav2"
using PT4;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace PT4Tasks
{
    public class MyTask : PT
    {
        /*
        ПРИМЕЧАНИЕ. Описанный вариант взаимодействия субъектов
        и наблюдателей можно реализовать с помощью ДЕЛЕГАТОВ
        и СОБЫТИЙ (см. примечание к предыдущему заданию
        в заготовке для языка C#).
        */

        public abstract class Subject
        {
            public abstract void Attach(Observer observ);
            public abstract void Detach(Observer observ);
            public abstract void SetInfo(string info);
        }

        // РЕАЛИЗУЙТЕ КЛАСС-ПОТОМОК ConcreteSubject
        public class ConcreteSubject :  Subject{
	    List<Observer> observers = new List<Observer>();
	//string message;
		public override void Attach(Observer observ) {
			observers.Add(observ);
		}
    	public override void Detach(Observer observ) {
    		observers.Remove(observ);
            Show("dett");
    		//iterator = observers.erase(iterator);iterator --;
		}
    	public override void SetInfo(string info) {
			//iterator = observers.begin();
            //for (int i = 0;i<observers.Count;i++) observers[i].OnInfo(this,info);
            foreach(var obs in observers.ToList()) obs.OnInfo(this,info);
            ////
		}/////////////////ПРОСКАКИВАЕТ СЛУДУЮЩИЙ ЕСЛИ УДАЛИЛ
};

        public abstract class Observer
        {
            public abstract void OnInfo(Subject sender, string info);
        }

        // РЕАЛИЗУЙТЕ КЛАСС-ПОТОМОК ConcreteObserver

        public class ConcreteObserver :  Observer
{
	public string log;
	public char detachInfo;
    public bool flag;

	public ConcreteObserver (char detachInfo){
		this.detachInfo = detachInfo;
		log = "";
	}
    public override void OnInfo(Subject sender, string info) {
    	log += info;
    	if (info[1] == detachInfo){
		Show("det");Show(info); Detach(sender);
		} 
	}
	public void Attach (Subject s){
		s.Attach(this);
	}
	public void Detach (Subject s){
		Show("det");
		s.Detach(this);
	}
	public string GetLog() {return log;}
};

        public static void Solve()
        {
            Task("OOP3Behav2");

            ConcreteSubject cs1 = new ConcreteSubject();
            ConcreteSubject cs2 = new ConcreteSubject();

            int n = GetInt(),k = GetInt();

            ConcreteObserver[] obs = new ConcreteObserver [n];

            for (int i=0;i<n;i++){
            	Show("init");
                int t = Convert.ToInt32('a')+ i;
                char tmp = (char)t;
            	obs[i] = new ConcreteObserver(tmp);
            	cs1.Attach(obs[i]);
            	cs2.Attach(obs[i]);
	        }

	        for (int i=0;i<k;i++){
	        	Show("in work");
	        	string tmp= GetString();
	        	if (tmp[0] == '1') cs1.SetInfo(tmp);
	        	else cs2.SetInfo(tmp);
	        }

            for (int i=0;i<n;i++){
            	Put(obs[i].GetLog());
	        }
    
        }
    }
}
